File: phi4_HMC/devel/prove pre-simulazioni/autocorrPlot.m
function autocorrPlot (nameFile,data1)

if nargin < 1 
    nameFile = 'provaHMC.txt';
    data1 = load(nameFile,'\t');
end
if nargin < 2 data1 = load(nameFile,'\t'); end
% Import data
%counter = data1(:,1);
m1 = data1(:,1);
%absMag = data1(:,3);
%M2 = data1(:,4);
%M4 = data1(:,5);

meanMag1 = mean(m1);

% Set iniziale grafico
figure()
axis square
title("autocorrPlot");
hold on;

N = length(m1);

nMax = N/10;
points = ones(nMax,1);

for t = 1:nMax
    gamma = 0;
    for i = 1:N-t
        gamma = gamma + (m1(i)-meanMag1)*(m1(i+t)-meanMag1);
    end
    points(t) = gamma/(N-t);
end

t = 1:nMax;
scatter(t,points,'.','r');
hold off;

end

File: autocorrPlot.py
import numpy as np
import matplotlib.pyplot as plt

def autocorrPlot (nameFile = 'provaHMC_naccu100.txt'):
    
    dataFile = open(nameFile,'r')
    #num_lines = np.sum(1 for line in dataFile)
    nData = 0
    data = np.array([])
    for i,line in enumerate(dataFile):
        if len(line.split()) == 0:
            continue
        if not line.startswith("%"):
            data = np.append(data,np.fromstring(line,sep=' \t '))
            nData += 1
    data = data.reshape((nData,-1))
    
    # Import data
    data = np.delete(data,0,1) # drop out the counters in the first column
    #m1 = data[:,0]
    #absMag = data[:,1]
    #M2 = data[:,2]
    #M4 = data[:,3]
    
    means = np.mean(data,axis = 0)
    #meanMag1 = np.mean(m1) = means[0]
    #meanAbsMag = np.mean(absMag) = means[1]
    #meanM2 = np.mean(M2) = means[2]
    #meanM4 = np.mean(M4) = means[3]

    for i in range(len(means)):
        # Set iniziale grafico
        fig = plt.figure(i+1)
        #ax = fig.add_subplot(111)
        #ax.set_aspect('equal', adjustable='box')
        if i == 0: titolo = "autocorrelation plot for $m$"
        elif i == 1: titolo = "autocorrelation plot for $|m|$"
        elif i == 2: titolo = "autocorrelation plot for $m^2$"
        elif i == 3: titolo = "autocorrelation plot for $m^4$"
        else: print('Neglect other observables')
        plt.title(titolo)
        
        N = nData
        nMax = int(N/10)
        points = np.ones(nMax)
        
        for t in range(nMax):
            gamma = 0
            for j in range(N-t):
                gamma = gamma + (data[j][i]-means[i])*(data[j+t][i]-means[i])
            points[t] = gamma/(N-t-1)

        t = np.arange(1,nMax+1)
        plt.scatter(t,points,c='r',marker='.')
        plt.show()

    

#autocorrPlot('phi4_HMC/main/DATI/ntraj_fixed/naccu1.txt')
